dev/test/Makefile

```diff
@@ -0,0 +1,164 @@
+endif
+
+# autodect a lot of various supports on current platform
+$(info ---------------------------------------------)
+
+NVCC := $(shell which nvcc 2>/dev/null)
+
+# Check and include cudnn if available
+# You can override the path to cudnn frontend by setting CUDNN_FRONTEND_PATH on the make command line
+# By default, we look for it in HOME/cudnn-frontend/include and ./cudnn-frontend/include
+# Refer to the README for cuDNN install instructions
+ifeq ($(USE_CUDNN), 1)
+  ifeq ($(shell [ -d $(HOME)/cudnn-frontend/include ] && echo "exists"), exists)
+    $(info ✓ cuDNN found, will run with flash-attention)
+    CUDNN_FRONTEND_PATH ?= $(HOME)/cudnn-frontend/include
+  else ifeq ($(shell [ -d cudnn-frontend/include ] && echo "exists"), exists)
+    $(info ✓ cuDNN found, will run with flash-attention)
+    CUDNN_FRONTEND_PATH ?= cudnn-frontend/include
+  else
+    $(error ✗ cuDNN not found. See the README for install instructions and the Makefile for hard-coded paths)
+  endif
+  NVCC_INCLUDES += -I$(CUDNN_FRONTEND_PATH)
+  NVCC_LDFLAGS += -lcudnn
+  NVCC_FLAGS += -DENABLE_CUDNN
+  NVCC_CUDNN = $(BUILD_DIR)/cudnn_att.o
+else
+  $(info → cuDNN is manually disabled by default, run make with `USE_CUDNN=1` to try to enable)
+endif
+
+# Check if OpenMP is available
+# This is done by attempting to compile an empty file with OpenMP flags
+# OpenMP makes the code a lot faster so I advise installing it
+# e.g. on MacOS: brew install libomp
+# e.g. on Ubuntu: sudo apt-get install libomp-dev
+# later, run the program by prepending the number of threads, e.g.: OMP_NUM_THREADS=8 ./gpt2
+# First, check if NO_OMP is set to 1, if not, proceed with the OpenMP checks
+ifeq ($(NO_OMP), 1)
+  $(info OpenMP is manually disabled)
+else
+  ifneq ($(OS), Windows_NT)
+    # Check for OpenMP support in GCC or Clang on Linux
+    ifeq ($(shell echo | $(CC) -fopenmp -x c -E - > /dev/null 2>&1; echo $$?), 0)
+      CFLAGS += -fopenmp -DOMP
+      LDLIBS += -lgomp
+      $(info ✓ OpenMP found)
+    else
+      $(info ✗ OpenMP not found)
+    endif
+  endif
+endif
+
+# Check if OpenMPI and NCCL are available, include them if so, for multi-GPU training
+ifeq ($(NO_MULTI_GPU), 1)
+  $(info → Multi-GPU (OpenMPI + NCCL) is manually disabled)
+else
+  ifeq ($(shell [ -d /usr/lib/x86_64-linux-gnu/openmpi/lib/ ] && [ -d /usr/lib/x86_64-linux-gnu/openmpi/include/ ] && echo "exists"), exists)
+    $(info ✓ OpenMPI found, OK to train with multiple GPUs)
+    NVCC_INCLUDES += -I/usr/lib/x86_64-linux-gnu/openmpi/include
+    NVCC_LDFLAGS += -L/usr/lib/x86_64-linux-gnu/openmpi/lib/
+    NVCC_LDLIBS += -lmpi -lnccl
+    NVCC_FLAGS += -DMULTI_GPU
+  else
+    $(info ✗ OpenMPI is not found, disabling multi-GPU support)
+    $(info ---> On Linux you can try install OpenMPI with `sudo apt install openmpi-bin openmpi-doc libopenmpi-dev`)
+  endif
+endif
+
+# Precision settings, default to bf16 but ability to override
+PRECISION ?= BF16
```

Comment:
should we actually define a default for CI?

Reply:
Good point - yes may be best to force a PRECISION on the command line so we always know what we're building.

Reply:
Added this explicit requirement for PRECISION being set.